File: src/main/java/com/fengsheng/skill/DuiZhengXiaYao.java
package com.fengsheng.skill;

import com.fengsheng.*;
import com.fengsheng.card.Card;
import com.fengsheng.card.PlayerAndCard;
import com.fengsheng.phase.FightPhaseIdle;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Errcode;
import com.fengsheng.protos.Role;
import com.google.protobuf.GeneratedMessageV3;
import org.apache.log4j.Logger;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.EnumMap;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

/**
 * 黄济仁技能【对症下药】：争夺阶段，你可以翻开此角色牌，然后摸三张牌，并且你可以展示两张含有相同颜色的手牌，然后从一名角色的情报区，弃置一张对应颜色情报。
 */
public class DuiZhengXiaYao extends AbstractSkill implements ActiveSkill {
    private static final Logger log = Logger.getLogger(DuiZhengXiaYao.class);

    @Override
    public SkillId getSkillId() {
        return SkillId.DUI_ZHENG_XIA_YAO;
    }

    @Override
    public void executeProtocol(Game g, Player r, GeneratedMessageV3 message) {
        if (!(g.getFsm() instanceof FightPhaseIdle fsm) || r != fsm.whoseFightTurn) {
            log.error("现在不是发动[对症下药]的时机");
            return;
        }
        if (r.isRoleFaceUp()) {
            log.error("你现在正面朝上，不能发动[对症下药]");
            return;
        }
        var pb = (Role.skill_dui_zheng_xia_yao_a_tos) message;
        if ((r instanceof HumanPlayer humanPlayer) && !humanPlayer.checkSeq(pb.getSeq())) {
            log.error("操作太晚了, required Seq: " + humanPlayer.getSeq() + ", actual Seq: " + pb.getSeq());
            return;
        }
        r.incrSeq();
        r.addSkillUseCount(getSkillId());
        g.playerSetRoleFaceUp(r, true);
        log.info(r + "发动了[对症下药]");
        r.draw(3);
        g.resolve(new executeDuiZhengXiaYaoA(fsm, r));
    }

    private record executeDuiZhengXiaYaoA(FightPhaseIdle fsm, Player r) implements WaitingFsm {
        private static final Logger log = Logger.getLogger(executeDuiZhengXiaYaoA.class);

        @Override
        public ResolveResult resolve() {
            final Game g = r.getGame();
            for (Player p : g.getPlayers()) {
                if (p instanceof HumanPlayer player) {
                    var builder = Role.skill_dui_zheng_xia_yao_a_toc.newBuilder();
                    builder.setPlayerId(player.getAlternativeLocation(r.location()));
                    builder.setWaitingSecond(15);
                    if (player == r) {
                        final int seq2 = player.getSeq();
                        builder.setSeq(seq2);
                        player.setTimeout(GameExecutor.post(g, () -> {
                            if (player.checkSeq(seq2)) {
                                g.tryContinueResolveProtocol(r, Role.skill_dui_zheng_xia_yao_b_tos.newBuilder()
                                        .setEnable(false).setSeq(seq2).build());
                            }
                        }, player.getWaitSeconds(builder.getWaitingSecond() + 2), TimeUnit.SECONDS));
                    }
                    player.send(builder.build());
                }
            }
            if (r instanceof RobotPlayer) {
                GameExecutor.post(g, () -> {
                    EnumMap<Common.color, List<Card>> cache = new EnumMap<>(Common.color.class);
                    cache.put(Common.color.Black, new ArrayList<>());
                    cache.put(Common.color.Red, new ArrayList<>());
                    cache.put(Common.color.Blue, new ArrayList<>());
                    for (Card card : r.getCards().values()) {
                        for (Common.color color : card.getColors())
                            cache.get(color).add(card);
                    }
                    for (var colorAndCards : cache.entrySet()) {
                        List<Card> cards = colorAndCards.getValue();
                        if (cards.size() >= 2 && findColorMessageCard(g, List.of(colorAndCards.getKey())) != null) {
                            var builder = Role.skill_dui_zheng_xia_yao_b_tos.newBuilder().setEnable(true);
                            for (Card card : cards) {
                                builder.addCardIds(card.getId());
                                if (builder.getCardIdsCount() >= 2)
                                    break;
                            }
                            g.tryContinueResolveProtocol(r, builder.build());
                            return;
                        }
                    }
                    g.tryContinueResolveProtocol(r, Role.skill_dui_zheng_xia_yao_b_tos.newBuilder().setEnable(false).build());
                }, 1, TimeUnit.SECONDS);
            }
            return null;
        }

        @Override
        public ResolveResult resolveProtocol(Player player, GeneratedMessageV3 message) {
            if (player != r) {
                log.error("不是你发技能的时机");
                return null;
            }
            if (!(message instanceof Role.skill_dui_zheng_xia_yao_b_tos pb)) {
                log.error("错误的协议");
                return null;
            }
            Game g = r.getGame();
            if (r instanceof HumanPlayer humanPlayer && !humanPlayer.checkSeq(pb.getSeq())) {
                log.error("操作太晚了, required Seq: " + humanPlayer.getSeq() + ", actual Seq: " + pb.getSeq());
                return null;
            }
            if (!pb.getEnable()) {
                r.incrSeq();
                for (Player p : g.getPlayers()) {
                    if (p instanceof HumanPlayer player1)
                        player1.send(Role.skill_dui_zheng_xia_yao_b_toc.newBuilder().setPlayerId(player1.getAlternativeLocation(r.location())).setEnable(false).build());
                }
                fsm.whoseFightTurn = fsm.inFrontOfWhom;
                return new ResolveResult(fsm, true);
            }
            if (pb.getCardIdsCount() != 2) {
                log.error("enable为true时必须要发两张牌");
                return null;
            }
            Card[] cards = new Card[2];
            for (int i = 0; i < cards.length; i++) {
                Card card = r.findCard(pb.getCardIds(i));
                if (card == null) {
                    log.error("没有这张卡");
                    return null;
                }
                cards[i] = card;
            }
            List<Common.color> colors = getSameColors(cards[0], cards[1]);
            if (colors.isEmpty()) {
                log.error("两张牌没有相同的颜色");
                return null;
            }
            PlayerAndCard playerAndCard = findColorMessageCard(g, colors);
            if (playerAndCard == null) {
                log.error("场上没有选择的颜色的情报牌");
                if (player instanceof HumanPlayer humanPlayer)
                    humanPlayer.send(Errcode.error_code_toc.newBuilder().setCode(Errcode.error_code.no_color_message_card).build());
                return null;
            }
            r.incrSeq();
            return new ResolveResult(new executeDuiZhengXiaYaoB(fsm, r, cards, colors, playerAndCard), true);
        }
    }

    private record executeDuiZhengXiaYaoB(FightPhaseIdle fsm, Player r, Card[] cards, List<Common.color> colors,
                                          PlayerAndCard defaultSelection) implements WaitingFsm {
        private static final Logger log = Logger.getLogger(executeDuiZhengXiaYaoB.class);

        @Override
        public ResolveResult resolve() {
            log.info(r + "展示了" + Arrays.toString(cards));
            Game g = r.getGame();
            for (Player p : g.getPlayers()) {
                if (p instanceof HumanPlayer player1) {
                    var builder = Role.skill_dui_zheng_xia_yao_b_toc.newBuilder().setEnable(true);
                    for (Card card : cards)
                        builder.addCards(card.toPbCard());
                    builder.setPlayerId(player1.getAlternativeLocation(r.location()));
                    builder.setWaitingSecond(15);
                    if (player1 == r) {
                        final int seq2 = player1.getSeq();
                        builder.setSeq(seq2);
                        player1.setTimeout(GameExecutor.post(g, () -> {
                            if (player1.checkSeq(seq2))
                                g.tryContinueResolveProtocol(r, Role.skill_dui_zheng_xia_yao_c_tos.newBuilder()
                                        .setTargetPlayerId(r.getAlternativeLocation(defaultSelection.player().location()))
                                        .setMessageCardId(defaultSelection.card().getId()).setSeq(seq2).build());
                        }, player1.getWaitSeconds(builder.getWaitingSecond() + 2), TimeUnit.SECONDS));
                    }
                    player1.send(builder.build());
                }
            }
            if (r instanceof RobotPlayer) {
                GameExecutor.post(g, () -> g.tryContinueResolveProtocol(r, Role.skill_dui_zheng_xia_yao_c_tos.newBuilder()
                        .setTargetPlayerId(r.getAlternativeLocation(defaultSelection.player().location()))
                        .setMessageCardId(defaultSelection.card().getId()).build()), 1, TimeUnit.SECONDS);
            }
            return null;
        }

        @Override
        public ResolveResult resolveProtocol(Player player, GeneratedMessageV3 message) {
            if (player != r) {
                log.error("不是你发技能的时机");
                return null;
            }
            if (!(message instanceof Role.skill_dui_zheng_xia_yao_c_tos pb)) {
                log.error("错误的协议");
                return null;
            }
            Game g = r.getGame();
            if (r instanceof HumanPlayer humanPlayer && !humanPlayer.checkSeq(pb.getSeq())) {
                log.error("操作太晚了, required Seq: " + humanPlayer.getSeq() + ", actual Seq: " + pb.getSeq());
                return null;
            }
            if (pb.getTargetPlayerId() < 0 || pb.getTargetPlayerId() >= g.getPlayers().length) {
                log.error("目标错误");
                return null;
            }
            Player target = g.getPlayers()[r.getAbstractLocation(pb.getTargetPlayerId())];
            if (!target.isAlive()) {
                log.error("目标已死亡");
                return null;
            }
            Card card = target.findMessageCard(pb.getMessageCardId());
            if (card == null) {
                log.error("没有这张牌");
                return null;
            }
            boolean contains = false;
            for (Common.color color : colors) {
                if (card.getColors().contains(color)) {
                    contains = true;
                    break;
                }
            }
            if (!contains) {
                log.error("选择的情报不含有指定的颜色");
                return null;
            }
            r.incrSeq();
            log.info(r + "弃掉了" + target + "面前的" + card);
            g.getDeck().discard(target.deleteMessageCard(card.getId()));
            for (Player p : g.getPlayers()) {
                if (p instanceof HumanPlayer player1) {
                    var builder = Role.skill_dui_zheng_xia_yao_c_toc.newBuilder();
                    builder.setPlayerId(player1.getAlternativeLocation(r.location()));
                    builder.setTargetPlayerId(player1.getAlternativeLocation(target.location()));
                    builder.setMessageCardId(card.getId());
                    player1.send(builder.build());
                }
            }
            fsm.whoseFightTurn = fsm.inFrontOfWhom;
            return new ResolveResult(fsm, true);
        }
    }

    private static List<Common.color> getSameColors(Card card1, Card card2) {
        List<Common.color> colors = new ArrayList<>();
        for (Common.color color : new Common.color[]{Common.color.Black, Common.color.Red, Common.color.Blue}) {
            if (card1.getColors().contains(color) && card2.getColors().contains(color))
                colors.add(color);
        }
        return colors;
    }

    private static PlayerAndCard findColorMessageCard(Game game, List<Common.color> colors) {
        for (Player player : game.getPlayers()) {
            for (Card card : player.getMessageCards().values()) {
                for (Common.color color : card.getColors()) {
                    if (colors.contains(color))
                        return new PlayerAndCard(player, card);
                }
            }
        }
        return null;
    }

    public static boolean ai(FightPhaseIdle e, final ActiveSkill skill) {
        Player player = e.whoseFightTurn;
        if (player.isRoleFaceUp())
            return false;
        int playerCount = player.getGame().getPlayers().length;
        List<PlayerAndCard> playerAndCards = new ArrayList<>();
        for (Player p : player.getGame().getPlayers()) {
            if (p.isAlive()) {
                for (Card c : p.getMessageCards().values())
                    playerAndCards.add(new PlayerAndCard(p, c));
            }
        }
        if (playerAndCards.size() < playerCount) return false;
        if (ThreadLocalRandom.current().nextInt(playerCount * playerCount) != 0)
            return false;
        GameExecutor.post(player.getGame(), () -> skill.executeProtocol(
                player.getGame(), player, Role.skill_dui_zheng_xia_yao_a_tos.getDefaultInstance()
        ), 2, TimeUnit.SECONDS);
        return true;
    }
}

File: src/main/kotlin/skill/DuiZhengXiaYao.kt
package com.fengsheng.skill

import com.fengsheng.*
import com.fengsheng.card.Card
import com.fengsheng.card.PlayerAndCard
import com.fengsheng.card.count
import com.fengsheng.card.filter
import com.fengsheng.phase.FightPhaseIdle
import com.fengsheng.protos.*
import com.fengsheng.protos.Common.color
import com.fengsheng.protos.Common.color.*
import com.fengsheng.protos.Role.*
import com.google.protobuf.GeneratedMessage
import org.apache.logging.log4j.kotlin.logger
import java.util.concurrent.TimeUnit

/**
 * 黄济仁技能【对症下药】：争夺阶段，你可以翻开此角色牌，然后摸三张牌，并且你可以展示两张含有相同颜色的手牌，然后从一名角色的情报区，弃置一张对应颜色情报。
 */
class DuiZhengXiaYao : ActiveSkill {
    override val skillId = SkillId.DUI_ZHENG_XIA_YAO

    override val isInitialSkill = true

    override fun canUse(fightPhase: FightPhaseIdle, r: Player): Boolean = !r.roleFaceUp

    override fun executeProtocol(g: Game, r: Player, message: GeneratedMessage) {
        val fsm = g.fsm as? FightPhaseIdle
        if (r !== fsm?.whoseFightTurn) {
            logger.error("现在不是发动[对症下药]的时机")
            r.sendErrorMessage("现在不是发动[对症下药]的时机")
            return
        }
        if (r.roleFaceUp) {
            logger.error("你现在正面朝上，不能发动[对症下药]")
            r.sendErrorMessage("你现在正面朝上，不能发动[对症下药]")
            return
        }
        val pb = message as skill_dui_zheng_xia_yao_a_tos
        if (r is HumanPlayer && !r.checkSeq(pb.seq)) {
            logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${pb.seq}")
            r.sendErrorMessage("操作太晚了")
            return
        }
        r.incrSeq()
        r.addSkillUseCount(skillId)
        g.playerSetRoleFaceUp(r, true)
        logger.info("${r}发动了[对症下药]")
        r.draw(3)
        g.resolve(ExecuteDuiZhengXiaYaoA(fsm, r))
    }

    private data class ExecuteDuiZhengXiaYaoA(val fsm: FightPhaseIdle, val r: Player) : WaitingFsm {
        override val whoseTurn: Player
            get() = fsm.whoseTurn

        override fun resolve(): ResolveResult? {
            val g = r.game!!
            g.players.send { p ->
                skillDuiZhengXiaYaoAToc {
                    playerId = p.getAlternativeLocation(r.location)
                    waitingSecond = Config.WaitSecond
                    if (p === r) {
                        val seq2 = p.seq
                        seq = seq2
                        p.timeout = GameExecutor.post(g, {
                            if (p.checkSeq(seq2)) {
                                g.tryContinueResolveProtocol(r, skillDuiZhengXiaYaoBTos {
                                    enable = false
                                    seq = seq2
                                })
                            }
                        }, p.getWaitSeconds(waitingSecond + 2).toLong(), TimeUnit.SECONDS)
                    }
                }
            }
            if (r is RobotPlayer) {
                GameExecutor.post(g, {
                    val messageCards = fsm.inFrontOfWhom.messageCards.toList()
                    fsm.inFrontOfWhom.messageCards.add(fsm.messageCard) // 先假设这个人已经获得了这张情报
                    var chooseColor = Black
                    var value = -1
                    for (c in listOf(Red, Blue, Black)) { // 对红、蓝、黑三种颜色进行判断
                        r.cards.count(c) >= 2 || continue // 如果这个颜色不够2张，则跳过
                        for (card in messageCards) { // 遍历目标角色面前的所有情报
                            c in card.colors || continue // 如果这张情报不含有这个颜色，则跳过
                            val v = r.calculateRemoveCardValue(fsm.whoseTurn, fsm.inFrontOfWhom, card) // 计算弃掉这张情报的价值
                            if (v > value) { // 如果这张情报的价值更高，则选这张
                                value = v
                                chooseColor = c
                            }
                        }
                    }
                    fsm.inFrontOfWhom.messageCards.removeLast() // 把刚才加上的那张情报移除
                    if (value < 0) { // 如果没有找到合适的情报，则不展示两张牌
                        g.tryContinueResolveProtocol(r, skillDuiZhengXiaYaoBTos { })
                        return@post
                    }
                    val cards = r.cards.filter(chooseColor).shuffled().take(2) // 随机选这个颜色的两张牌
                    g.tryContinueResolveProtocol(r, skillDuiZhengXiaYaoBTos {
                        enable = true
                        cards.forEach { cardIds.add(it.id) }
                    })
                }, 3, TimeUnit.SECONDS)
            }
            return null
        }

        override fun resolveProtocol(player: Player, message: GeneratedMessage): ResolveResult? {
            if (player !== r) {
                logger.error("不是你发技能的时机")
                player.sendErrorMessage("不是你发技能的时机")
                return null
            }
            if (message !is skill_dui_zheng_xia_yao_b_tos) {
                logger.error("错误的协议")
                player.sendErrorMessage("错误的协议")
                return null
            }
            val g = r.game!!
            if (r is HumanPlayer && !r.checkSeq(message.seq)) {
                logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${message.seq}")
                r.sendErrorMessage("操作太晚了")
                return null
            }
            if (!message.enable) {
                r.incrSeq()
                g.players.send {
                    skillDuiZhengXiaYaoBToc {
                        playerId = it.getAlternativeLocation(r.location)
                        enable = false
                    }
                }
                return ResolveResult(fsm.copy(whoseFightTurn = fsm.inFrontOfWhom), true)
            }
            if (message.cardIdsCount != 2) {
                logger.error("enable为true时必须要发两张牌")
                player.sendErrorMessage("不足两张牌")
                return null
            }
            val cards = List(2) { i ->
                val card = r.findCard(message.getCardIds(i))
                if (card == null) {
                    logger.error("没有这张卡")
                    player.sendErrorMessage("没有这张卡")
                    return null
                }
                card
            }
            val colors = getSameColors(cards[0], cards[1])
            if (colors.isEmpty()) {
                logger.error("两张牌没有相同的颜色")
                player.sendErrorMessage("两张牌没有相同的颜色")
                return null
            }
            val playerAndCard = r.findColorMessageCard(colors)
            if (playerAndCard == null) {
                logger.error("场上没有选择的颜色的情报牌")
                player.sendErrorMessage("场上没有选择的颜色的情报牌")
                return null
            }
            r.incrSeq()
            return ResolveResult(ExecuteDuiZhengXiaYaoB(fsm, r, cards, colors, playerAndCard), true)
        }
    }

    private data class ExecuteDuiZhengXiaYaoB(
        val fsm: FightPhaseIdle,
        val r: Player,
        val cards: List<Card>,
        val colors: List<color>,
        val defaultSelection: PlayerAndCard
    ) : WaitingFsm {
        override val whoseTurn: Player
            get() = fsm.whoseTurn

        override fun resolve(): ResolveResult? {
            logger.info("${r}展示了${cards.joinToString()}")
            val g = r.game!!
            g.players.send { p ->
                skillDuiZhengXiaYaoBToc {
                    playerId = p.getAlternativeLocation(r.location)
                    enable = true
                    this@ExecuteDuiZhengXiaYaoB.cards.forEach { cards.add(it.toPbCard()) }
                    waitingSecond = Config.WaitSecond
                    if (p === r) {
                        val seq2 = p.seq
                        seq = seq2
                        p.timeout = GameExecutor.post(g, {
                            if (p.checkSeq(seq2)) {
                                g.tryContinueResolveProtocol(r, skillDuiZhengXiaYaoCTos {
                                    targetPlayerId = r.getAlternativeLocation(defaultSelection.player.location)
                                    messageCardId = defaultSelection.card.id
                                    seq = seq2
                                })
                            }
                        }, p.getWaitSeconds(waitingSecond + 2).toLong(), TimeUnit.SECONDS)
                    }
                }
            }
            if (r is RobotPlayer) {
                GameExecutor.post(g, {
                    fsm.inFrontOfWhom.messageCards.add(fsm.messageCard) // 先假设这个人已经获得了这张情报
                    var playerAndCard: PlayerAndCard? = null
                    var value = Int.MIN_VALUE
                    for (p in g.players) {
                        p!!.alive || continue
                        for (card in p.messageCards.sortedBy { it.isBlack() }) { // 遍历所有玩家面前的所有情报
                            // 先遍历不带黑颜色的情报，这样就不会出现优先弃掉对方的真黑双色情报的问题了
                            card !== fsm.messageCard || continue // 如果遍历到了前面假设获得的那张情报，则跳过
                            card.colors.any { it in colors } || continue // 如果这张情报不含有这个颜色，则跳过
                            val v = r.calculateRemoveCardValue(fsm.whoseTurn, p, card) // 计算弃掉这张情报的价值
                            if (v > value) { // 如果这张情报的价值更高，则选这张
                                value = v
                                playerAndCard = PlayerAndCard(p, card)
                            }
                        }
                    }
                    fsm.inFrontOfWhom.messageCards.removeLast() // 把刚才加上的那张情报移除
                    playerAndCard!! // 从逻辑上来说，必定能找到一个合适的情报，否则上一条协议会拦截
                    g.tryContinueResolveProtocol(r, skillDuiZhengXiaYaoCTos {
                        targetPlayerId = r.getAlternativeLocation(playerAndCard.player.location)
                        messageCardId = playerAndCard.card.id
                    })
                }, 3, TimeUnit.SECONDS)
            }
            return null
        }

        override fun resolveProtocol(player: Player, message: GeneratedMessage): ResolveResult? {
            if (player !== r) {
                logger.error("不是你发技能的时机")
                player.sendErrorMessage("不是你发技能的时机")
                return null
            }
            if (message !is skill_dui_zheng_xia_yao_c_tos) {
                logger.error("错误的协议")
                player.sendErrorMessage("错误的协议")
                return null
            }
            val g = r.game!!
            if (r is HumanPlayer && !r.checkSeq(message.seq)) {
                logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${message.seq}")
                r.sendErrorMessage("操作太晚了")
                return null
            }
            if (message.targetPlayerId < 0 || message.targetPlayerId >= g.players.size) {
                logger.error("目标错误")
                player.sendErrorMessage("目标错误")
                return null
            }
            val target = g.players[r.getAbstractLocation(message.targetPlayerId)]!!
            if (!target.alive) {
                logger.error("目标已死亡")
                player.sendErrorMessage("目标已死亡")
                return null
            }
            val card = target.findMessageCard(message.messageCardId)
            if (card == null) {
                logger.error("没有这张牌")
                player.sendErrorMessage("没有这张牌")
                return null
            }
            var contains = false
            for (color in colors) {
                if (card.colors.contains(color)) {
                    contains = true
                    break
                }
            }
            if (!contains) {
                logger.error("选择的情报不含有指定的颜色")
                player.sendErrorMessage("选择的情报不含有指定的颜色")
                return null
            }
            r.incrSeq()
            logger.info("${r}弃掉了${target}面前的$card")
            g.deck.discard(target.deleteMessageCard(card.id)!!)
            g.players.send {
                skillDuiZhengXiaYaoCToc {
                    playerId = it.getAlternativeLocation(r.location)
                    targetPlayerId = it.getAlternativeLocation(target.location)
                    messageCardId = card.id
                }
            }
            return ResolveResult(fsm.copy(whoseFightTurn = fsm.inFrontOfWhom), true)
        }
    }

    companion object {
        private fun getSameColors(card1: Card, card2: Card) =
            listOf(Black, Red, Blue).filter { it in card1.colors && it in card2.colors }

        private fun Player.findColorMessageCard(colors: List<color?>): PlayerAndCard? {
            for (player in game!!.sortedFrom(game!!.players, location)) {
                for (card in player.messageCards) {
                    for (color in card.colors) {
                        if (color in colors) return PlayerAndCard(player, card)
                    }
                }
            }
            return null
        }

        fun ai(e: FightPhaseIdle, skill: ActiveSkill): Boolean {
            val player = e.whoseFightTurn
            val target = e.inFrontOfWhom
            val g = player.game!!
            !player.roleFaceUp || return false
            if (g.players.any {
                    it!!.isPartnerOrSelf(player) && it.willWin(e.whoseTurn, e.inFrontOfWhom, e.messageCard)
                }) return false
            g.players.any {
                it!!.isEnemy(player) && it.willWin(e.whoseTurn, e.inFrontOfWhom, e.messageCard)
            } || target.isPartnerOrSelf(player) && target.willDie(e.messageCard) || return false
            GameExecutor.post(player.game!!, {
                skill.executeProtocol(player.game!!, player, skillDuiZhengXiaYaoATos { })
            }, 3, TimeUnit.SECONDS)
            return true
        }
    }
}
